pkg/resources/data.go

```diff
@@ -469,6 +470,53 @@ func (d *TemplateData) GetOpenVPNServerPort() (int32, error) {
 	return service.Spec.Ports[0].NodePort, nil
 }
 
+// GetAPIServerAlternateNames returns the alternate names for the apiserver certificate from the corresponding services.
+// This method ensures that if multiple hostnames or IPs have been assigned to the API server service or front-loadbalancer service, then all
+// of them are included in the certificate.
+func (d *TemplateData) GetAPIServerAlternateNames() (*certutil.AltNames, error) {
+	// Get all the loadbalancer Ingresses from the API server service.
+	DNSNames := []string{}
+	IPs := []net.IP{}
+	service := &corev1.Service{}
+	if err := d.client.Get(d.ctx, types.NamespacedName{Namespace: d.cluster.Status.NamespaceName, Name: ApiserverServiceName}, service); err != nil {
```

Comment:
I'm not super happy with this. The templateData struct was meant to _hold_ data, and every function in it that dynamically fetches stuff is IHMO "bad". The fact that we keep a context in the template data is already an ugly hack, just so the individual reconciler functions do not all have to carry a context.

But in reality, I think this is an evil hack. Reconcilers were meant to be stateless and we're cheating by wrapping a context in here, so that the reconcilers do not notice the "effort" that goes on behind the scenes when they call .GetSomething(). We should strive to not rely further on this evil, evil, nasty embedded context.

I would prefer if the TemplateData is initialized with the necessary data at some point and then just holds it. There could be a function on the template data builder that takes a context and client and fills in a .apiserverAltNames field.

Reply:
The reason I kept this as a `method` that fetches is because these values can change and are not static. Anyways will refactor.

Reply:
Updated, PTAL and let me know what you think ;)

Reply:
Much nicer now, thanks.

Reply:
Nice but wouldn't work :D

Reply:
Unfortunately, had to revert my last commit since it would never work. We cannot populate the API Server Alternative names in the template data since the process of creating the services(api-server and front-loadbalancer) _starts_ after the template data is ready. Before that, the services wouldn't exist, and it'll keep failing. 

I don't see much of a point in waiting for the consequent reconciliations after the services have been created and then assuming that if the services are there, then populate the template data with apiServerAlternativeNames otherwise keep that field empty and ignore errors regarding services not existing. I think it adds more toil since we won't get the correct secret for API server certificates in the first reconciliation and then some subsequent reconciliation will regenerate the certificate secret with all the alt names.

Looking at how the templateData methods are already being used; they are already doing what I'm trying to do in this PR. https://github.com/kubermatic/kubermatic/blob/main/pkg/resources/data.go#L253 or search for `func (d *TemplateData) Get` in that file.

But please, I'm not trying to push you opinion away, if we still can find an alternative then I'm more than happy to implement it :)

Reply:
It also seems like an anti-pattern to me, template data is used to render and create the resources. It itself shouldn't then depend on those rendered resources to populate it's own data.

Reply:
I disagree that this is an anti-pattern. The whole idea of a struct to hold all necessary data for the oject creator functions is that they become independent of the cluster state and that you can test and render them standalone. In most of the KKP codebase we first gather all information necessary and then give that to the object creators, i.e. most have no context parameter and no kube client. Similar to rendering a Go template, where I also wouldn't give the template a database client, but fetch all necessary data first and hand it in.

> Looking at how the templateData methods are already being used; they are already doing what I'm trying to do in this PR. https://github.com/kubermatic/kubermatic/blob/main/pkg/resources/data.go#L253 or search for func (d *TemplateData) Get in that file.

Absolutely, and all of these cases are bad and should not serve as an example for new code. If it wasn't such a PITA to rewrite the codebase to get rid of the implicit context/kubeclient in the templateData, I'd do it in a heartbeat :-D

> I don't see much of a point in waiting for the consequent reconciliations

I disagree to a point. If the components in the cluster namespace depend on the service's status, we have to wait or reconcile later again. At the moment this PR has a race condition because you create the Service and then immediately get the Service's status, probably before the CCM had time to provision an LB. So effectively once the LB Service changes, you will re-reconcile anyway and extend the serving cert anyway.

A sequence like

```
reconciling.ReconcileServices(data)
data.additionalAltNames = GetAPIServerAlternateNames(ctx, client)
reconciling.ReconcileCertificates(data)
reconciling.Reconcile....(data)
```

this will have the same behaviour (first reconciliation will probably not have all altNames yet, subsequent ones will have them), but still get rid of the implicit context in the templateData.

We *could* wait and re-reconcile if the GetAPIServerAlternateNames() somehow indicates a "not yet ready, please wait" depending on the chosen expose strategy. This would mean before the LB is ready, most of the control plane will not be deployed. As in

```
reconciling.ReconcileServices(data)
data.additionalAltNames, ready, err := GetAPIServerAlternateNames(ctx, client)
if err != nil {
  return err
}
if !ready {
  return nil // we should have a Watch that gets triggered on the neccessary updates anyway
}
reconciling.ReconcileCertificates(data)
reconciling.Reconcile....(data)
```

I could see how that's nice but also how that's bad.

> It also seems like an anti-pattern to me, template data is used to render and create the resources. It itself shouldn't then depend on those rendered resources to populate its own data.

That sounds like you're agreeing with me, cause in this PR right now, templateData implicitly depends on some kube objects existing 😁

Reply:
> That sounds like you're agreeing with me, cause in this PR right now, templateData implicitly depends on some kube objects existing

Not really. It's an anti-pattern because you are populating it with static data within a structure. The methods that are being used and what I'm introducing are just using static data(context, client etc) to fetch data "dynamically." That dynamic data will not exist in the cluster so we are not really populating anything the template itself.

Reply:
> Looking at how the templateData methods are already being used; they are already doing what I'm trying to do in this PR. https://github.com/kubermatic/kubermatic/blob/main/pkg/resources/data.go#L253 or search for func (d *TemplateData) Get in that file.

I'd suggest sticking to how it's already working in this current PR. We can open a separate issue if we want to refactor this code, but TBH, it needs a proper discussion in sig-cluster-management. Since I find this wrong as well:

```
reconciling.ReconcileServices(data)
data.additionalAltNames = GetAPIServerAlternateNames(ctx, client)
reconciling.ReconcileCertificates(data)
reconciling.Reconcile....(data)
```

`data.additionalAltNames` additionalAltNames is private and shouldn't be exposed like this. The setters for it `WithAPIServerAlternativeNames` can only be used when building a new TemplateData. Yes, we can expose a new method that lets us modify this field whenever we want to, but it already sounds so bad. I liked the aspect where once the template data is built by the template builder, the static data https://github.com/kubermatic/kubermatic/blob/release/v2.25/pkg/resources/data.go#L64-L97 that it encapsulates doesn't change.

But I agree with your point that we need to find some consistency in how this pattern is used. Unfortunately, I don't want to introduce changes in this PR that are not **just** related to this PR; these are more general changes.

Reply:
I've created https://github.com/kubermatic/kubermatic/issues/13563 so that we can do whatever the refactoring is at a much broader scope/level than just in this PR. That should unblock this PR from further review :)